controllers/machine_controller_phases.go

```diff
@@ -78,6 +78,13 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, m *clusterv1.
 	obj, err := external.Get(r.Client, ref, m.Namespace)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
+			if m.Status.InfrastructureReady && m.Status.GetTypedPhase() == clusterv1.MachinePhaseRunning {
+				// Infra object went missing after the machine was up and running
```

Comment:
The resource that went missing could be either bootstrap or infra

Reply:
I thought the error status should only be set when infra resource is missing. A missing bootstrap object should not result in an error, isn't it?

Reply:
Bootstrap.ConfigRef is optional, someone could explicitly supply Bootstrap.Data instead.

Reply:
I was referring to the case where it goes

1. Have bootstrap + infra, machine boots & is infra-ready
1. Delete bootstrap

Reply:
@ncdc even in that case as long as `Bootstrap.Data` is already set, do we actually care?

Reply:
@ncdc I am also thinking the same as @detiber

Reply:
There was someone in the community who mentioned they wanted to react to KubeadmConfig changes in their environment, so in that case, it would matter

Reply:
@ncdc possibly, but I wouldn't consider that a "terminal error"

Reply:
Fair